File: python_extensions/libs/python/test/const_aware.py
import const_aware_mod
import unittest
import pickle

const_aware_mod.Example.__reduce__ = lambda self: (const_aware_mod.Example, (self.value_prop,))

class TestConstAware(unittest.TestCase):

    def setUp(self):
        self.owner = const_aware_mod.Owner()

    def checkNonConst(self, x):
        self.assertEqual(type(x), const_aware_mod.Example)        
        self.assert_(x.const_method())
        self.assert_(x.non_const_method())
        self.assert_(self.owner.accept_by_value(x))
        self.assert_(self.owner.accept_by_const_value(x))
        self.assert_(self.owner.accept_by_reference(x))
        self.assert_(self.owner.accept_by_const_reference(x))
        self.assert_(self.owner.accept_by_shared_ptr(x))
        self.assert_(self.owner.accept_by_const_shared_ptr(x))
        x.value_prop = 0
        self.assertEqual(x.value_prop, 0)
        self.assertEqual(x.value_ro, 0)
        self.assertEqual(x.value_rw, 0)
        x.value_prop = 1
        self.assertEqual(x.value_prop, 1)
        self.assertEqual(x.value_ro, 1)
        self.assertEqual(x.value_rw, 1)
        x.value_rw = 0
        self.assertEqual(x.value_prop, 0)
        self.assertEqual(x.value_ro, 0)
        self.assertEqual(x.value_rw, 0)
        def set_value_ro(v):
            x.value_ro = v
        self.assertRaises(AttributeError, set_value_ro, 1)

    def checkConst(self, x):
        self.assertEqual(type(x), const_aware_mod.Example.__const_proxy__)
        self.assert_(x.const_method())
        self.assertFalse(hasattr(x,"non_const_method"))
        self.assert_(self.owner.accept_by_value(x))
        self.assert_(self.owner.accept_by_const_value(x))
        self.assertRaises(Exception, self.owner.accept_by_reference, x)
        self.assert_(self.owner.accept_by_const_reference(x))
        self.assertRaises(Exception, self.owner.accept_by_shared_ptr, x)
        self.assert_(self.owner.accept_by_const_shared_ptr(x))
        self.assertEqual(x.value_prop, 0)
        self.assertEqual(x.value_ro, 0)
        self.assertEqual(x.value_rw, 0)
        def set_value_prop(v):
            x.value_prop = v
        def set_value_rw(v):
            x.value_rw = v
        def set_value_ro(v):
            x.value_ro = v
        self.assertRaises(AttributeError, set_value_prop, 1)
        self.assertRaises(AttributeError, set_value_rw, 1)
        self.assertRaises(AttributeError, set_value_ro, 1)

    def testByValue(self):
        by_value = self.owner.by_value()
        by_const_value = self.owner.by_const_value()
        self.assertNotEqual(by_value.address, by_const_value.address)
        self.checkNonConst(by_value)
        self.checkConst(by_const_value)

    def testByReference(self):
        by_reference = self.owner.by_reference()
        by_const_reference = self.owner.by_const_reference()
        self.assertEqual(by_reference.address, by_const_reference.address)
        self.checkNonConst(by_reference)
        self.checkConst(by_const_reference)

    def testBySharedPtr(self):
        by_shared_ptr = self.owner.by_shared_ptr();
        by_const_shared_ptr = self.owner.by_const_shared_ptr();
        self.assertEqual(by_shared_ptr.address, by_const_shared_ptr.address)
        self.checkNonConst(by_shared_ptr)
        self.checkConst(by_const_shared_ptr)

    def testConstruction(self):
        original = self.owner.by_value()
        proxy = const_aware_mod.Example.__const_proxy__(original)
        self.assertEqual(original.address, original.address)
        self.checkConst(proxy)
        original_copy = const_aware_mod.Example(original)
        proxy_copy = const_aware_mod.Example(proxy)
        self.assertNotEqual(original.address, original_copy.address)
        self.assertNotEqual(proxy.address, proxy_copy.address)
        self.checkNonConst(original_copy)
        self.checkNonConst(proxy_copy)

    def testPickle(self):
        original = self.owner.by_value()
        original.value_prop = 3
        original_pickled = pickle.dumps(original)
        original_loaded = pickle.loads(original_pickled)
        self.assertEqual(original_loaded.value_prop, original.value_prop)
        proxy = const_aware_mod.Example.__const_proxy__(original)
        proxy_pickled = pickle.dumps(proxy)
        proxy_loaded = pickle.loads(proxy_pickled)
        self.assertEqual(proxy_loaded.value_prop, proxy.value_prop)
        self.checkNonConst(proxy_loaded)

if __name__=="__main__":
    unittest.main()

File: python_extensions/libs/python/test/const_aware.cpp
#include <boost/python.hpp>
#include <boost/python/const_aware.hpp>
#include <boost/shared_ptr.hpp>

namespace bp = boost::python;

class Example {
public:
    Example() : value(0) {}

    std::size_t get_address() const { return std::size_t(this); }

    bool non_const_method() { return true; }
    bool const_method() const { return true; }

    int value;

    int get_value() const { return value; }
    void set_value(int v) { value = v; }
};

class Owner {
public:
    Example by_value() const { return *_example; }
    Example const by_const_value() const { return *_example; }
    Example & by_reference() { return *_example; }
    Example const & by_const_reference() const { return *_example; }
    boost::shared_ptr<Example> by_shared_ptr() { return _example; }
    boost::shared_ptr<Example const> by_const_shared_ptr() const { return _example; }

    bool accept_by_value(Example x) { return true; }
    bool accept_by_const_value(Example const x) { return true; }
    bool accept_by_reference(Example & x) { return true; }
    bool accept_by_const_reference(Example const & x) { return true; }
    bool accept_by_shared_ptr(boost::shared_ptr<Example> const & x) { return true; }
    bool accept_by_const_shared_ptr(boost::shared_ptr<Example const> const & x) { return true; }

    explicit Owner() : _example(new Example()) {}
private:
    boost::shared_ptr<Example> _example;
};

static void export_module() {

    bp::class_<Example> pyExample("Example");
    pyExample.def(bp::init<Example const &>());
        
    bp::make_const_aware(pyExample, "FrozenExample")
        .add_property("address", &Example::get_address)
        .def("non_const_method", &Example::non_const_method)
        .def("const_method", &Example::const_method)
        .add_property("value_prop", &Example::get_value, &Example::set_value)
        .def_readonly("value_ro", &Example::value)
        .def_readwrite("value_rw", &Example::value)
        .enable_shared_ptr();
        ;

    bp::class_<Owner>("Owner")
        .def("by_value", &Owner::by_value)
        .def("by_const_value", &Owner::by_const_value, bp::as_const<>())
        .def("by_reference", &Owner::by_reference, bp::return_internal_reference<>())
        .def("by_const_reference", &Owner::by_const_reference, 
             bp::as_const< bp::return_internal_reference<> >())
        .def("by_shared_ptr", &Owner::by_shared_ptr)
        .def("by_const_shared_ptr", &Owner::by_const_shared_ptr)
        .def("accept_by_value", &Owner::accept_by_value)
        .def("accept_by_const_value", &Owner::accept_by_const_value)
        .def("accept_by_reference", &Owner::accept_by_reference)
        .def("accept_by_const_reference", &Owner::accept_by_const_reference)
        .def("accept_by_shared_ptr", &Owner::accept_by_shared_ptr)
        .def("accept_by_const_shared_ptr", &Owner::accept_by_const_shared_ptr)
        ;

}

BOOST_PYTHON_MODULE(const_aware) {
    export_module();
}
